akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -411,16 +411,23 @@ private[http] object HttpServerBluePrint {
           val response = grab(httpResponseIn)
           val requestStart = openRequests.head
           openRequests = openRequests.tail
-          val isEarlyResponse = messageEndPending && openRequests.isEmpty
-          if (isEarlyResponse && response.status.isSuccess)
+          val (isEarlyResponse, close) = if (settings.proxyMode) {
+            val early = (messageEndPending && openRequests.isEmpty) && !requestStart.expect100Continue
+            val close = requestStart.closeRequested || (isClosed(requestParsingIn) && openRequests.isEmpty)
+            (early, close)
+          } else {
+            val early = messageEndPending && openRequests.isEmpty
+            val close = requestStart.closeRequested ||
+              (requestStart.expect100Continue && oneHundredContinueResponsePending) ||
+              (isClosed(requestParsingIn) && openRequests.isEmpty) || early
+            (early, close)
+          }
+
+          if (isEarlyResponse && response.status.isSuccess && !requestStart.expect100Continue)
             log.warning(
-              "Sending an 2xx 'early' response before end of request was received... " +
+              "Sending a {} 'early' response before end of request was received... " +
```

Comment:
I think we should fix it properly by not looking for `status.isSuccess` but by checking if it is a subclass of `StatusCode.Success`.